src/uu/dd/src/dd.rs

```diff
@@ -401,42 +395,58 @@ where
         let output_thread = thread::spawn(gen_prog_updater(rx, i.settings.status));
         let mut progress_as_secs = 0;
 
-        // Create a common buffer with a capacity of the block size.
-        // This is the max size needed.
-        let mut buf = vec![BUF_INIT_BYTE; bsize];
-
-        // The main read/write loop.
-        //
-        // Each iteration reads blocks from the input and writes
-        // blocks to this output. Read/write statistics are updated on
-        // each iteration and cumulative statistics are reported to
-        // the progress reporting thread.
-        while below_count_limit(&i.settings.count, &rstat, &wstat) {
-            // Read a block from the input then write the block to the output.
+        // Optimization: if no blocks are to be written, then don't
+        // bother allocating any buffers.
+        let will_write_at_least_one_block = match i.settings.count {
+            Some(Num::Blocks(n)) if n > 0 => true,
+            Some(Num::Bytes(n)) if n > 0 => true,
+            None => true,
+            _ => false,
+        };
```

Comment:
Nit, but another way to write this could be:
```rust
if let Some(Num::Blocks(0) | Num::Bytes(0)) =  i.settings.count {
    return Ok()
}
```

This is a bit more succinct and the early return helps with readability in my opinion.

Reply:
I agree with the sentiment, though I think we don't want to `return` here since we still want to print the final read/write statistics (the block of code starting at line 453 below). Let me see if there's some straightforward way to incorporate this idea to improve readability.

Reply:
Okay, I've updated it to use this idea, thank you.